### PR TITLE
SUSTAIN Wear-time Reorientation

### PR DESCRIPTION
**Pull request by DMegaritis**
## Summary

Adds validated wear-time detection, and sensor reorientation correction algorithms.

### New Algorithms
- Three wear-time detection methods (`WtdMegaritisSignal`, `WtdMegaritisXGBoost`, `WtdMegaritisCNN`) with benchmarks
- `ReorientationMethodDM` for correcting IMU misorientation (axis permutations/inversions)

### Documentation
- Comprehensive guides for algorithm selection, coordinate systems, and aggregation workflows
- API references and practical examples for wear-time and reorientation
- Updated Sphinx configuration for new modules

### Pipeline
- Weartime detection integrated into pipeline
- Optional reorientation correction integrated into pipeline
- Weekly DMA aggregation examples using pipeline-detected wear-time

### CODE REVIEW

File: src/mobgap/re_orientation/_re_orientation.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+
+    else:
+        where_grav = None
+        where_grav_points = None
+        family = None
+
+    return where_grav, where_grav_points, family
+
+
+def _swap_is_ml(corrected: pd.DataFrame, original: pd.DataFrame) -> pd.DataFrame:
```
Comment (AKuederle) [inline src/mobgap/re_orientation/_re_orientation.py:213]: Wait! You can't do this! You can not flip sensor axis independently for IMUs! You always need to rotate. never flip. Otherwise rotations are not defined properly in the new coordinate system.

Maybe migrate the rotate and flip helpers from gaitmap to mobgap as well (note that we have a folder for all functions we directly copied from gaitmap)

File: src/mobgap/re_orientation/_re_orientation.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    return _flip_ap(_flip_ml(data))
+
+
+# Filter parameters
+FILTER_LOWCUT = 0.5  # Hz
+FILTER_HIGHCUT = 2.5  # Hz
+FILTER_ORDER = 100
+
+
+def _bandpass_filter(signal_data: np.ndarray, lowcut: float, highcut: float, fs: int, order: int) -> np.ndarray:
```
Comment (AKuederle) [inline src/mobgap/re_orientation/_re_orientation.py]: We already have filter implementations in mobgap./ Reuse

File: src/mobgap/re_orientation/_re_orientation.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        f, pxx_is = signal.welch(acc_is_filt, fs=fs, nperseg=nperseg)
+        stride_mask = (f >= 0.5) & (f <= 2.5)
+        if not np.any(stride_mask):
+            return 0.0
+        is_power = pxx_is[stride_mask]
+        phase = np.angle(cxy[stride_mask])
+        if is_power.sum() > 0:
+            return float(np.average(phase, weights=is_power))
+    except ValueError:
+        pass
```
Comment (AKuederle) [inline src/mobgap/re_orientation/_re_orientation.py]: ? we can not just silently ignore a Value Error. Why does the ValueError happen?

File: src/mobgap/re_orientation/_re_orientation.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+    return 0.0
+
+
+def _apply_ml_ap_correction(
```
Comment (AKuederle) [inline src/mobgap/re_orientation/_re_orientation.py:304]: I think it would be much cleaner, if you align the algorithm around a rotation matrix. You identify two known directions, then create a rotation matrix freom it and store it. Much cleaner no need to remember distinct families

File: src/mobgap/weartime/utils/ml_feature_extraction.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
+"""Helper functions for machine learning-based wear-time detection."""
+
+import numpy as np
+import pandas as pd
+from scipy.signal import welch
+
+
+def rolling_window_indices(n_samples: int, win_samples: int, step: int) -> tuple[int, int]:
```
Comment (AKuederle) [inline src/mobgap/weartime/utils/ml_feature_extraction.py:8]: I think we already have moving window utils that are optimized. If not in mobgap, check gaitmap

File: src/mobgap/weartime/__init__.py
Comment (AKuederle) [inline src/mobgap/weartime/__init__.py:1]: Is it worth adding all 3 algorithms? When should you use which? How good are they? How are they validated?

File: src/mobgap/weartime/_wtd_megaritis_cnn.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        version: Literal["cnn", "cnn_lstm"] = "cnn_lstm",
+        position: Literal["lowback"] = "lowback",
+    ) -> None:
+        self.window_sec = window_sec
+        self.overlap = overlap
+        self.version = version
+        self.position = position
+
+        # Load model based on version
+        if self.version == "cnn":
```
Comment (AKuederle) [inline src/mobgap/weartime/_wtd_megaritis_cnn.py]: See the LR/R algorithm how to properly load models from file without violating tpcp "no modifications in init rules"

File: src/mobgap/weartime/_wtd_megaritis_cnn.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            model_file = files("mobgap.weartime.production_models").joinpath("cnn_lowback_model.keras")
+        else:  # cnn_lstm
+            model_file = files("mobgap.weartime.production_models").joinpath("cnn_lstm_lowback_model.keras")
+
+        self.model = keras.models.load_model(model_file)
+
+    def __getstate__(self) -> dict:
```
Comment (AKuederle) [inline src/mobgap/weartime/_wtd_megaritis_cnn.py:128]: Puuuh.... this will break stuff... I will have a look how to handle this properly

File: src/mobgap/weartime/_wtd_megaritis_cnn.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        sample-level wear-time segments using majority voting and biomechanical
+        post-processing rules.
+
+        Parameters
+        ----------
+        %(detect_para)s
+
+        %(detect_return)s
+
+        Notes
```
Comment (AKuederle) [inline src/mobgap/weartime/_wtd_megaritis_cnn.py:159]: Should be documented in class not in method

File: src/mobgap/weartime/_wtd_megaritis_cnn.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+        1. Majority voting: Each sample receives votes from overlapping windows
+        2. Short bout removal: Wear bouts <15s are removed (too short to don/doff)
+        3. Confidence filter: Wear bouts <20min require ≥90%% vote agreement
+           (boundary bouts at start/end of data are exempt)
+        4. Gap merging: Non-wear gaps <15s between wear periods are merged
+        """
+        self.data = data
+        self.sampling_rate_hz = sampling_rate_hz
+        self.data_length = len(data)
```
Comment (AKuederle) [inline src/mobgap/weartime/_wtd_megaritis_cnn.py]: Don't store it. Either it is a result or you should not store it

File: src/mobgap/weartime/_wtd_megaritis_cnn.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        step = int(win_samples * (1 - self.overlap))
+        n_samples = self.data_length
+
+        # Required columns for CNN (6 channels: 3 acc + 3 gyr)
+        required_cols = ["acc_is", "acc_ml", "acc_pa", "gyr_is", "gyr_ml", "gyr_pa"]
+
+        all_predictions = []
+
+        # Extract windows and predict
+        for start, end in rolling_window_indices(n_samples, win_samples, step):
```
Comment (AKuederle) [inline src/mobgap/weartime/_wtd_megaritis_cnn.py]: Is this the recommended approach for keras to do inference on moving window? This sounds extremly inefficient. You should at least batch load into GPU

File: src/mobgap/weartime/_wtd_megaritis_cnn.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+            win = self.data.iloc[start:end]
+
+            # Extract raw IMU window and standardize per-window
+            x_window = win[required_cols].to_numpy().astype(np.float32)
+
+            # Per-window standardization (matching training preprocessing)
+            x_mean = x_window.mean(axis=0)  # shape: (6,)
```
Comment (AKuederle) [inline src/mobgap/weartime/_wtd_megaritis_cnn.py]: This works? Per window standardisation is always quite dangerous and basically makes all "scale" base parameters relativly useless.

File: tests/tests_weartime/test_wtd_megaritis_cnn.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+class TestMetaWtdMegaritisCNN(TestAlgorithmMixin):
+    __test__ = True
+    ALGORITHM_CLASS = WtdMegaritisCNN
+
+    @pytest.fixture
+    def after_action_instance(self):
+        # Use random data to avoid edge cases
+        return self.ALGORITHM_CLASS().detect(
+            pd.DataFrame(np.random.randn(1000, 6), columns=BF_SENSOR_COLS), sampling_rate_hz=100.0
```
Comment (AKuederle) [inline tests/tests_weartime/test_wtd_megaritis_cnn.py:27]: Fix the Seed for things like this, that tests don't randomly fail one day

File: pyproject.toml [hunk 13]
```diff
@@ -35,6 +35,10 @@ dependencies = [
     "numba>=0.60.0",
     "python-dotenv>=1.0.1",
     "seaborn>=0.12.0",
+    # Additions for weartime and reorientatoin
+    "pyentrp>=2.1.0",
+    "tensorflow>=2.18.0,<2.22; sys_platform != 'win32' and python_version < '3.14'",
```
Comment (AKuederle) [inline pyproject.toml:40]: The way you wrote the code, assumes that these are optionial dependencies

**Review by AKuederle — CHANGES_REQUESTED** [review]
High level review (so did not look into the actual functionality or correctness of any of the algorithms).

Structure looks good, but we need to clarify a couple of fundamental things tto decide how we best intgrate this -> see email

File: src/mobgap/weartime/_wtd_megaritis_cnn.py (hunk 6 above)
Comment (DMegaritis) [reply src/mobgap/weartime/_wtd_megaritis_cnn.py]: I have moved the model loading outside __init__ to follow tpcp best practices, it's now loaded via a cached property

File: src/mobgap/weartime/_wtd_megaritis_cnn.py (hunk 7 above)
Comment (DMegaritis) [reply src/mobgap/weartime/_wtd_megaritis_cnn.py:128]: Fixed since the model is no longer loaded in __init__ :)

File: src/mobgap/weartime/_wtd_megaritis_cnn.py (hunk 8 above)
Comment (DMegaritis) [reply src/mobgap/weartime/_wtd_megaritis_cnn.py:159]: Removed from class, thank you.

File: src/mobgap/weartime/_wtd_megaritis_cnn.py (hunk 9 above)
Comment (DMegaritis) [reply src/mobgap/weartime/_wtd_megaritis_cnn.py]: Thanks, I am not storing data_length now.

File: src/mobgap/weartime/_wtd_megaritis_cnn.py (hunk 10 above)
Comment (DMegaritis) [reply src/mobgap/weartime/_wtd_megaritis_cnn.py]: Since then, I have updated this to perform batch processing, following Keras best practices for efficient GPU utilisation

File: src/mobgap/weartime/_wtd_megaritis_cnn.py (hunk 11 above)
Comment (DMegaritis) [reply src/mobgap/weartime/_wtd_megaritis_cnn.py]: Yes, this works correctly! Per-window standardisation was used during model training. The model was specifically designed to detect wear-time based on movement patterns and temporal structure (rotation frequencies, waveform shapes) rather than absolute signal magnitudes. 

PS: Per-window vs global scaling vs no scalling was evaluated as a hyperparameter during training and the per-window showed superior performance.

File: tests/tests_weartime/test_wtd_megaritis_cnn.py (hunk 12 above)
Comment (DMegaritis) [reply tests/tests_weartime/test_wtd_megaritis_cnn.py:27]: Thanks, seed added.

File: pyproject.toml (hunk 13 above)
Comment (DMegaritis) [reply pyproject.toml:40]: This is intentional. I was getting Windows CI build failures with TensorFlow in and there is no support in Python ≥3.14. The platform/version conditions ensure MobGap installs successfully on all systems. 
Please let me know if you think this should be handled differently.

File: src/mobgap/re_orientation/_re_orientation.py (hunk 2 above)
Comment (DMegaritis) [reply src/mobgap/re_orientation/_re_orientation.py]: Updated to mobgap's filter. Thank you.

File: src/mobgap/re_orientation/_re_orientation.py (hunk 3 above)
Comment (DMegaritis) [reply src/mobgap/re_orientation/_re_orientation.py]: Thanks, the ValueError occurred when walking bouts were shorter than the filter's padlen. I have removed the silent error handling and now explicitly check data length before filtering (short bouts return 0.0).

File: src/mobgap/weartime/utils/ml_feature_extraction.py (hunk 5 above)
Comment (DMegaritis) [reply src/mobgap/weartime/utils/ml_feature_extraction.py:8]: I replaced my function with sliding_window_view from mobgap, thank you

File: src/mobgap/weartime/__init__.py
Comment (DMegaritis) [reply src/mobgap/weartime/__init__.py:1]: The algorithms were developed and validated as part of a SUSTAIN/EFPIA-funded project. Full validation results and methodology will be published in an upcoming paper (currently in preparation). A novel dataset with ground-truth wear/non-wear labels was collected and used for algorithm development and validation alongside additional datasets.

I've included error, speed, and selection guidance in docs/guides/weartime_model_selection.md. 

Brief summary:
- WtdMegaritisCNN: Most accurate (~1.3 min daily error) and fastest (3× faster pipeline runtime), but requires TensorFlow (struggles with some Windows environments and Python ≥3.14)
- WtdMegaritisSignal: Cross-platform fallback with no dependencies, good performance (~15 min daily error), but less accurate than CNN
- WtdMegaritisXGBoost: Middle-ground ML option (~2.5 min daily error) for environments where TensorFlow isn't supported but XGBoost is available, though computationally expensive (5-17 min processing time)

All three algorithms represent novel methodologies using accelerometer and gyroscope data to detect wear-like movement patterns, significantly outperforming previous threshold-based device-movement approaches from the literature. The documentation clearly prioritises CNN, then signal processing, with XGBoost as an intermediate option. Since all algorithms pass tests and serve distinct deployment scenarios (TensorFlow-supported vs dependency-minimal vs XGBoost-only environments), I believe including all three provides comprehensive coverage without introducing confusion.

File: src/mobgap/re_orientation/_re_orientation.py (hunk 1 above)
Comment (DMegaritis) [reply src/mobgap/re_orientation/_re_orientation.py:213]: The operations performed are reflections and permutations, not pure rotations. After testing, I found that some corrections can be approximated with rotation combinations, but this approach is less transparent and introduces unnecessary complexity.
The current flip/swap operations are deterministic, and I have validated them on real-world data, novel datasets with devices places in actual wrong orientations, and synthetic data, and produce physically plausible sensor orientations that align with the body frame. While intermediate steps may temporarily create implausible axis configurations (e.g., single-axis flips), the final output for each of the eight potential orientations is guaranteed to be a valid right-handed coordinate system correctly aligned to anatomical axes.
I've updated the coordinate system documentation and Figure to show the mathematical transformations explicitly (coordinate_system.md). Each transformation is expressed as the minimal set of operations required to restore the reference frame from that specific misorientation.
If you have a reliable method to achieve these reflections and permutations using only scipy.Rotation (which rejects improper rotations with determinant ≠ +1), I'm happy to reconsider. Otherwise, the current approach is both mathematically correct and practically validated. I am sharing the link to the technical report via email.

File: src/mobgap/re_orientation/_re_orientation.py (hunk 4 above)
Comment (DMegaritis) [reply src/mobgap/re_orientation/_re_orientation.py:304]: Families are a presentation framework for the user, the algorithm itself detects gravity direction and movement phase to determine corrections dynamically. The family classification is for diagnostics/reporting, not the core correction logic.

**Comment by DMegaritis** [comment]
Many thanks for your review @AKuederle, changes implemented :)

**Comment by AKuederle** [comment]
This is now split into two dedicated PRs: #237 #236. Will close this one. @DMegaritis  If the two dedicated PRs are missing anything, please migrate the missing parts over
